module3/lessons/Week4/UpdatingResources.md

```diff
@@ -0,0 +1,219 @@
+    context.Movies.Add(movie);
+    context.SaveChanges();
+
+    var formData = new Dictionary<string, string>
+    {
+        { "Title", "Goofy" },
+        { "Genre", "Documentary" }
+    };
+
+    // Act
+    var response = await client.PostAsync($"/movies/{movie.Id}", new FormUrlEncodedContent(formData));
+    var html = await response.Content.ReadAsStringAsync();
+
+    // Assert
+    response.EnsureSuccessStatusCode();
+    Assert.Contains("Goofy", html);
+    Assert.Contains("Documentary", html);
+    Assert.DoesNotContain("Comedy", html);
+}
+```
+
+</section>
+
+### the Order of Our Tests
+
+Before we implement our controller action, we need to address one aspect of testing - the order and timing of each of our tests.  By default, xUnit will run our tests randomly, and as quickly as possible (some tests even run at the same time!).  Generally, this is a good thing - we want our tests to be individually robust, and not rely on any _other_ test to run successfully. In this case, though, we need to control our test runs a bit more.
+
+If left to their default settings, we could have two tests running at the same time (concurrently) that would put conflicting data into our test database.  xUnit provides a simple way to group tests that need to be run strictly one at a time.  We will use an attribute to group all of our Controller tests into one collection:
+
+```c#
+// MoviesControllerTests.cs
+
+namespace MvcMovie.FeatureTests
+{
+    [Collection("Controller Tests")]
+    public class MovieControllerTests : IClassFixture<WebApplicationFactory<Program>>
+    {
+        ...
+```
+
+Any test class in this collection will be run one at a time, reducing the chance for a database conflict.
+
+<section class='call-to-action' markdown='1'>
+
+**With a Partner**  
+Implement the controller action that we need to update records in our database.
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### One Solution
+```c#
+[HttpPost]
+[Route("/Movies/{id:int}")]
+public IActionResult Update(Movie movie)
+{
+    _context.Movies.Update(movie);
+    _context.SaveChanges();
+
+    return RedirectToAction("show", new { id = movie.Id });
+}
+```
+
+</section>
+
+## TDD Updating related resource
+**Code-Along**  
+As a class, we are going to write tests, and implement code to satisfy these user-stories:
+
+```
+As a User
+When I visit a Movie Show page
+Then I see links to edit each review
+```
+
+```
+As a User
+When I visit 'Movies/1/Reviews/1/edit'
+Then I see the name of the Movie and,
+Then I see a form to edit the review
+    And the review fields are pre-populated with their current value
+```
+
+```
+As a User
+When I Submit the review-edit form
+Then I am redirected to the Movie Show Page
+    And I see my updated review
+```
+
+
+<section class='instructor-notes' markdown='1'>
+
+Completed code can be found on the update-resources branch of the [MvcMovieStarter](https://github.com/turingschool-examples/MvcMovieStarter)
```

Comment:
I just checked out your branch and it looks like you forgot to push the updates for editing a review. I can edit a movie but not a review.

Reply:
Also, what do you think about always being super explicit with naming path parameters.

`[Route("/Movies/{movieId:int}/reviews/{reviewId:int}")]`

instead of 

``[Route("/Movies/{movieId:int}/reviews/{id:int}")]``

Reply:
I'm good with being super specific :)

Reply:
Are you getting an error when you edit a review?  I don't have any additional changes... and I'm able to edit a review.

I am going to push up a change so that there is an 'edit' button on each review from the movies/:id/reviews page - that is a usability miss for sure 😆

Reply:
Oh... fun times... actually .NET does not really like it when I use a more specific id.  When I update my controller action to:

```
[HttpPost]
[Route("/Movies/{movieId:int}/reviews/{reviewId:int}")]
public IActionResult Update(int movieId, int reviewId, Review review)
{
    _context.Reviews.Update(review);
    _context.SaveChanges();

    return RedirectToAction("index", new { movieId = movieId });
}
```

.NET does not automatically associate the reviewId with the review object.  I think the .NET 'magic' is that it will assume that an `id` param is the id of that controller's resource.